src/ObjectUtils/CoffFile.cpp

```diff
@@ -258,34 +315,131 @@ ErrorMessageOr<ModuleSymbols> CoffFileImpl::LoadDebugSymbols() {
-  if (object_file_->getSymbolTable() != 0 && object_file_->getNumberOfSymbols() != 0) {
-    return false;
+bool CoffFileImpl::HasDebugSymbols() const {
+  if (object_file_->getSymbolTable() != 0) {
+    for (const auto& symbol_ref : object_file_->symbols()) {
+      llvm::Expected<llvm::object::SymbolRef::Type> type = symbol_ref.getType();
+      if (type && type.get() != llvm::object::SymbolRef::ST_Function) {
+        return true;
+      }
+    }
   }
 
   const std::unique_ptr<llvm::DWARFContext> dwarf_context =
       llvm::DWARFContext::create(*object_file_);
   if (dwarf_context == nullptr) {
-    ORBIT_ERROR("Could not create DWARF context for \"%s\"", file_path_.string());
-    return true;
+    return false;
   }
+  for (const auto& compile_unit : dwarf_context->compile_units()) {
+    for (uint32_t die_index = 0; die_index < compile_unit->getNumDIEs(); ++die_index) {
+      llvm::DWARFDie full_die = compile_unit->getDIEAtIndex(die_index);
+      if (!full_die.isSubprogramDIE()) {
+        continue;
+      }
 
-  for (const auto& info_section : dwarf_context->compile_units()) {
-    for (uint32_t index = 0; index < info_section->getNumDIEs(); ++index) {
-      llvm::DWARFDie full_die = info_section->getDIEAtIndex(index);
-      if (!full_die.isSubprogramDIE()) continue;
-
-      uint64_t low_pc;
-      uint64_t high_pc;
-      uint64_t unused_section_index;
-      if (full_die.getLowAndHighPC(low_pc, high_pc, unused_section_index)) {
-        return false;
+      uint64_t low_pc{};
+      uint64_t high_pc{};
+      uint64_t unused_section_index{};
+      if (full_die.getLowAndHighPC(low_pc, high_pc, unused_section_index) && low_pc != 0) {
+        return true;
       }
     }
   }
-  return true;
+  return false;
+}
+
+// LLVM doesn't provide a simple way to access the RUNTIME_FUNCTIONs, so we have to do it ourselves.
+// Even llvm-objdump has to do that:
+// https://github.com/llvm-mirror/llvm/blob/master/tools/llvm-objdump/COFFDump.cpp
+ErrorMessageOr<llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>> CoffFileImpl::GetRuntimeFunctions() {
+  const llvm::object::data_directory* exception_table_data_dir =
+      object_file_->getDataDirectory(llvm::COFF::EXCEPTION_TABLE);
+  if (exception_table_data_dir == nullptr) {
+    return ErrorMessage{"Unable to read Exception Table: No corresponding Data Directory."};
+  }
+
+  llvm::ArrayRef<uint8_t> exception_table_bytes;
+  if (object_file_->getRvaAndSizeAsBytes(exception_table_data_dir->RelativeVirtualAddress,
+                                         exception_table_data_dir->Size, exception_table_bytes)) {
+    return ErrorMessage{"Unable to read Exception Table."};
+  }
+
+  if (exception_table_bytes.size() % sizeof(llvm::Win64EH::RuntimeFunction) != 0) {
+    return ErrorMessage{"Unable to read Exception Table: Unexpected size."};
+  }
+
+  return llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>{
+      reinterpret_cast<const llvm::Win64EH::RuntimeFunction*>(exception_table_bytes.data()),
```

Comment:
Nit: Shall we use absl::bitcast?

Reply:
Yep, it is fine. Though we mostly use bitcast also for those. So I was wandering if we should remain consistent. No need to change this though.